### PR TITLE
Add AES encryption and decryption

### PR DESCRIPTION
**Pull request by torkelrogstad**
Test suite might be a bit on the light side, and I'd also prefer to flesh out error cases in both encrypting and decrypting. But TBH I'm not quite sure how to do that. 

Edit: 

Looking at Bitcoin Core [source](https://github.com/bitcoin/bitcoin/blob/fa6180188b8ab89af97860e6497716405a48bab6/src/crypto/aes.h) it looks like they also use AES with CBC block cipher

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                                           salt = salt,
+                                           iterationCount = ITERATIONS,
+                                           derivedKeyLength = KEY_SIZE)
+    new SecretKeySpec(tmp.getEncoded, "AES")
+  }
+}
+
+sealed trait AesEncryptionException extends Exception
+
+sealed trait AesDecryptionException extends Exception
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala:164]: `RuntimeException`?

File: core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                new IvParameterSpec(initializationVector.toArray))
+    cipher
+  }
+
+  /**
+    * Decrypts the provided data
+    */
+  def decrypt(
+      encrypted: AesEncryptedData,
+      password: String): Either[AesDecryptionException, ByteVector] = {
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala]: We should really consider making a type for passwords that wraps a String so that we never accidentally pass a password anywhere it isn't expected.

File: core/src/main/scala/org/bitcoins/core/crypto/PBKDF2.scala [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+  * @define keyStretch Derives the provided value and salt to a secret key
+  *                    using the PBKDF2 key derivation function
+  *
+  * Utilities related to the PBKDF2 key derivation function
+  */
+object PBKDF2 {
+
+  private val PSEUDO_RANDOM_FUNCTION = "PBKDF2WithHmacSHA512"
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/crypto/PBKDF2.scala:20]: Could you add a comment with a link to where this comes from?

File: core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+                                           derivedKeyLength = KEY_SIZE)
+    new SecretKeySpec(tmp.getEncoded, "AES")
+  }
+}
+
+sealed trait AesEncryptionException extends Exception
+
+sealed trait AesDecryptionException extends Exception
+
+case object BadPasswordException
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala]: Let's create a wrapper object for these? I.e. 

```scala
object AESException { 
   final case object BadPasswordException ...
   
   final case object EmptyPasswordException ...
}

File: core-test/src/test/scala/org/bitcoins/core/crypto/AesCryptTest.scala [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    forAll(NumberGenerator.bytevector) { bytes =>
+      val encrypted = AesCrypt.encryptExc(bytes, password)
+      val decryptedE = AesCrypt.decrypt(encrypted, badPassword)
+      decryptedE match {
+        case Right(decrypted) =>
+          assert(decrypted != bytes)
+        case Left(exc) => assert(exc == BadPasswordException)
+
+      }
+      succeed
```
Comment (Christewart) [inline core-test/src/test/scala/org/bitcoins/core/crypto/AesCryptTest.scala]: why is this needed?

File: core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    }
+  }
+
+  private def encryptionCipher(secret: SecretKey): Cipher = {
+    val cipher = getCipher
+    cipher.init(Cipher.ENCRYPT_MODE, secret)
+    cipher
+  }
+
+  lazy private val secureRandom: SecureRandom = new SecureRandom
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala]: i'm always leary of caching PRNG, let's not do this.

File: core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+  /**
+    * Generates a random hex string of length 30
+    */
+  private def getSalt: String = {
+    val stringBuilder = new StringBuilder
+    (0 until 30).foreach { _ =>
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala]: i'd prefer this 

```scala
    val array = Array[Byte](32)
    val rand = new SecureRandom(array)
    rand.nextBytes(array)
    require(array.exists(_ != 0), s"Array for salt must be initialized")
    ByteVector(array).toHex
```

I'm not sure what the optimal salt size is, the SO question you link to says it needs to be 8 bytes. The question is almost 10 years old though, so those security practices may be out dated.

Comment (torkelrogstad) [reply core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala]: Is there any reason we want to provide `array` as seed to the `SecureRandom`?

File: core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala (hunk 1 above)
Comment (torkelrogstad) [reply core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala:164]:
```
illegal inheritance; superclass BadPaddingException
[error]  is not a subclass of the superclass RuntimeException
[error]  of the mixin trait AesDecryptionException
[error]       with AesDecryptionException
```

Do you know how to fix that?

File: core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+  private def encryptionCipher(secret: SecretKey): Cipher = {
+    val cipher = getCipher
+    cipher.init(Cipher.ENCRYPT_MODE, secret)
+    cipher
+  }
+
+  private def getSecureRandom: SecureRandom = new SecureRandom
+
+  /**
+    * Generates a random hex string of length 30
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala]: commment fix

File: core-test/src/test/scala/org/bitcoins/core/crypto/AesCryptTest.scala [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+import org.bitcoins.testkit.util.BitcoinSUnitTest
+import scodec.bits.HexStringSyntax
+
+class AesCryptTest extends BitcoinSUnitTest {
+  behavior of "AesEncrypt"
+
+  val password = AesPassword("PASSWORD")
+  val badPassword = AesPassword("BAD_PASSWORD")
+  val emptyPassword = AesPassword("")
+
```
Comment (Christewart) [inline core-test/src/test/scala/org/bitcoins/core/crypto/AesCryptTest.scala:14]: We are missing a couple of hard coded test vectors with expected ecrypted output and then the corresponding decrypted output. These are really useful for sanity tests. Property tests can be misleading some times as you can implement the encrypt and decrypt functionality in just the right way so that your tests pass :-)
